components/object_store/src/disk_cache.rs

```diff
@@ -445,83 +657,123 @@ impl ObjectStore for DiskCacheStore {
+            if let Some(bytes) = self.cache.get_data(&filename, &range_in_file).await {
+                return Ok(bytes);
             }
+            // This page is missing from the disk cache, let's fetch it from the
+            // underlying store and insert it to the disk cache.
+            let aligned_bytes = self.fetch_and_cache_data(location, &aligned_range).await?;
+
+            // Allocate a new buffer instead of the `aligned_bytes` to avoid memory
+            // overhead.
+            let mut bytes_buf = BytesMut::with_capacity(range.len());
+            bytes_buf.extend_from_slice(
+                &aligned_bytes[(range.start - aligned_start)..(range.end - aligned_start)],
+            );
+            return Ok(bytes_buf.freeze());
         }
 
-        for range in missing_ranges {
-            let range_start = range.start;
-            let cache_key = Self::cache_key(location, &range);
-            // TODO: we should use get_ranges here.
-            let bytes = self.underlying_store.get_range(location, range).await?;
-            self.cache.insert(cache_key, bytes.clone()).await;
-            ranged_bytes.insert(range_start, bytes);
-        }
+        // The queried range involves multiple ranges.
+        let mut paged_bytes: Vec<Option<Bytes>> = vec![None; num_pages];
+        let mut num_missing_pages = 0;
+        {
+            let mut page_start = aligned_start;
+            let mut page_idx = 0;
+            while page_idx < num_pages {
+                let page_end = (page_start + self.page_size).min(file_size);
+                let range_in_file = {
+                    let real_start = page_start.max(range.start);
+                    let real_end = page_end.min(range.end);
+                    (real_start - page_start)..(real_end - page_start)
+                };
+                let filename = Self::page_filename(location, &(page_start..page_end));
+                if let Some(bytes) = self.cache.get_data(&filename, &range_in_file).await {
+                    paged_bytes[page_idx] = Some(bytes);
+                } else {
+                    num_missing_pages += 1;
+                }
 
-        // we get all bytes for each aligned_range, organize real bytes
-        // fast path
-        if ranged_bytes.len() == 1 {
-            let (range_start, bytes) = ranged_bytes.pop_first().unwrap();
-            let result = bytes.slice((range.start - range_start)..(range.end - range_start));
-            return Ok(result);
+                page_start += self.page_size;
+                page_idx += 1;
+            }
         }
 
-        // there are multiple aligned ranges for one request, such as
-        // range = [3, 33), page_size = 16, then aligned ranges will be
-        // [0, 16), [16, 32), [32, 48)
-        // we need to combine those ranged bytes to get final result bytes
+        // Fetch all the missing pages from the underlying store.
```

Comment:
I think original comments are helpful to understand logic here.

Reply:
Agreed.